keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,560 @@
+	key, quit := cjc.queue.Get()
+	if quit {
+		return false
+	}
+	defer cjc.queue.Done(key)
+	if err := cjc.sync(key.(string)); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing CronJobController %v, requeuing: %v", key.(string), err))
+		cjc.queue.AddRateLimited(key)
+	} else {
+		cjc.queue.Forget(key)
+	}
+	return true
+}
+
+```
+
+#### Multiple workers
+
+We also propose to have multiple workers controller by a flag similar to [statefulset controller](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/apps.go#L65). The default would be set to 5 similar to [statefulset](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/config/v1alpha1/defaults.go#L34)
+
+#### Handling Cron aspect
+
+To detect which CronJob has met its schedule and need to create Jobs we need to implement a timer component. We shall implement a Heap based timer algorithm. We will introduce a separate queue with the [`DelayingInterface`](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L37) that implements heap based single shot api [`AddAfter`](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L150). Every time we process an entry from this queue, we will add it back to the queue to simulate a periodic timer.
+
+For further reading:
+1. [Reinventing timer wheel](https://lwn.net/Articles/646950/)
+2. [Hashed and hierarchical timer wheel](http://www.cs.columbia.edu/~nahum/w6998/papers/sosp87-timing-wheels.pdf)
+3. [Golang timers in multi-cpu systems](https://github.com/golang/go/commit/76f4fd8a5251b4f63ea14a3c1e2fe2e78eb74f81)
+4. [Go timerwheel](https://github.com/RussellLuo/timingwheel)
+
+### Metrics
+
+We propose to add metrics that could expose the performance and health of the controller including and not limited to:
+- Skew (actualJobCreationTime-expectedJobCreationTime) - [Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)
+- Job failures - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- Job successes - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- API server errors - [Gauge](https://prometheus.io/docs/concepts/metric_types/#counter)
+- Job scheduling latency (podCreationTime - jobCreationTime) - [Histogram](https://prometheus.io/docs/concepts/metric_types/#histogram)
+- Queue depth (pending CronJob and Job entries in the Queue) - [Gauge](https://prometheus.io/docs/concepts/metric_types/#gauge)
+
+Queue depth can be surfaced from existing controller framework.
+
+### Add .status.lastSuccessfulTime
+
+[#issue/75674](https://github.com/kubernetes/kubernetes/issues/75674)
+Add `lastSuccessfulTime` to `.status` that tracks the last time the job completed successfully. This will augment the `lastScheduledTime` available in the `.status` in the v1beta1 api. Potential use is in monitoring (e.g. fire an alert if lastSuccessfulTime is more than X ago).
```

Comment:
Will `.status.lastSuccessfulTime` always record the last time a Job completes successfully? For example, if the CronJob allows concurrent Job, and an old Job completes after the new one, `.status.lastSuccessfulTime` will still record the completion time of the older Job, right?

Reply:
yes, b/c the other job (with `Forbid`) won't be allowed to run, with `Replace` or `Allow` it will always be the last job's time.